op-service/bgpo/oracle.go

```diff
@@ -0,0 +1,404 @@
+// to newHeads and track blob base fees.
+func NewBlobGasPriceOracle(ctx context.Context, rpcClient client.RPC, chainConfig *params.ChainConfig, log log.Logger, config *BlobGasPriceOracleConfig) *BlobGasPriceOracle {
+	defaultConfig := DefaultBlobGasPriceOracleConfig()
+	if config == nil {
+		config = defaultConfig
+	}
+	if config.PricesCacheSize <= 0 {
+		config.PricesCacheSize = defaultConfig.PricesCacheSize
+	}
+	if config.BlockCacheSize <= 0 {
+		config.BlockCacheSize = defaultConfig.BlockCacheSize
+	}
+	if config.MaxBlocks <= 0 {
+		config.MaxBlocks = defaultConfig.MaxBlocks
+	}
+	if config.Percentile <= 0 || config.Percentile > 100 {
+		config.Percentile = defaultConfig.Percentile
+	}
+
+	logger := log.With("module", "bgpo")
+
+	pollClient := client.NewPollingClient(ctx, logger, rpcClient, client.WithPollRate(250*time.Millisecond))
+
+	oracleCtx, cancel := context.WithCancel(ctx)
+	return &BlobGasPriceOracle{
+		client:                 pollClient,
+		chainConfig:            chainConfig,
+		log:                    log.With("module", "bgpo"),
+		prices:                 caching.NewLRUCache[uint64, *big.Int](config.Metrics, "bgpo_prices", config.PricesCacheSize),
+		blobTxsPriorityFeeCaps: caching.NewLRUCache[uint64, []*big.Int](config.Metrics, "bgpo_tips", config.BlockCacheSize),
+		maxBlocks:              config.MaxBlocks,
+		percentile:             config.Percentile,
+		ctx:                    oracleCtx,
+		cancel:                 cancel,
+	}
+}
+
+// Start begins subscribing to newHeads and processing headers.
+// Before subscribing, it pre-populates the cache with the last MaxBlocks blocks.
+// This method blocks until the context is canceled or an error occurs.
+func (o *BlobGasPriceOracle) Start() error {
+	// Pre-populate cache with recent blocks before subscribing
+	if err := o.prePopulateCache(); err != nil {
+		o.log.Warn("Failed to pre-populate cache, continuing anyway", "err", err)
+	}
+
+	headers := make(chan *types.Header, 10)
+
+	doSubscribe := func(ch chan<- *types.Header) (ethereum.Subscription, error) {
+		return o.client.Subscribe(o.ctx, "eth", ch, "newHeads")
+	}
+
+	sub, err := doSubscribe(headers)
+	if err != nil {
+		return err
+	}
+	o.sub = sub
+
+	o.log.Info("Blob gas price oracle started, subscribed to newHeads")
+
+	// Process headers as they arrive
+	for {
+		select {
+		case header := <-headers:
+			if err := o.processHeader(header); err != nil {
+				o.log.Error("Error processing header", "err", err, "block", header.Number.Uint64())
+			}
+		case err := <-sub.Err():
+			if err != nil {
+				o.log.Error("Subscription error", "err", err)
+				return err
+			}
+			return nil
+		case <-o.ctx.Done():
+			o.log.Info("Blob gas price oracle context canceled")
+			return nil
+		}
+	}
+}
+
+// prePopulateCache fetches and processes the last MaxBlocks blocks to pre-populate the cache.
+func (o *BlobGasPriceOracle) prePopulateCache() error {
+	ctx, cancel := context.WithTimeout(o.ctx, 30*time.Second)
```

Comment:
This timeout seems a little too long IMO. Ideally we could use some shared network timeout constant defined somewhere?